routing/astar_router.cpp

```diff
@@ -0,0 +1,190 @@
+    return IRouter::EndPointNotFound;
+
+  FSegId const start(startEdge.GetFeatureId().m_index, startEdge.GetSegId());
+  FSegId const finish(finishEdge.GetFeatureId().m_index, finishEdge.GetSegId());
+
+  string const country = m_countryFileFn(startPoint);
+  MwmSet::MwmId const mwmId(m_index.GetMwmIdByCountryFile(platform::CountryFile(country)));
+  IndexGraph graph(
+      CreateGeometry(m_index, mwmId, m_vehicleModelFactory->GetVehicleModelForCountry(country)));
+
+  if (!LoadIndex(mwmId, graph))
+    return IRouter::RouteFileNotExist;
+
+  AStarProgress progress(0, 100);
+
+  auto onVisitJunctionFn = [&delegate, &progress, &graph](JointId const & from,
+                                                          JointId const & /*finish*/) {
+    m2::PointD const & point = graph.GetPoint(from);
+    auto const lastValue = progress.GetLastValue();
+    auto const newValue = progress.GetProgressForDirectedAlgo(point);
+    if (newValue - lastValue > kProgressInterval)
+      delegate.OnProgress(newValue);
+    delegate.OnPointCheck(point);
+  };
+
+  AStarAlgorithm<IndexGraph> algorithm;
+
+  RoutingResult<JointId> routingResult;
+  AStarAlgorithm<IndexGraph>::Result const resultCode =
+      algorithm.FindPathBidirectional(graph, graph.InsertJoint(start), graph.InsertJoint(finish),
+                                      routingResult, delegate, onVisitJunctionFn);
+
+  switch (resultCode)
+  {
+  case AStarAlgorithm<IndexGraph>::Result::NoPath: return IRouter::RouteNotFound;
+  case AStarAlgorithm<IndexGraph>::Result::Cancelled: return IRouter::Cancelled;
+  case AStarAlgorithm<IndexGraph>::Result::OK:
+    vector<Junction> path = ConvertToJunctions(graph, routingResult.path);
+    ReconstructRoute(m_directionsEngine.get(), *m_roadGraph, delegate, path, route);
+    return IRouter::NoError;
+  }
+}
+
+bool AStarRouter::FindClosestEdge(m2::PointD const & point, Edge & closestEdge) const
+{
+  vector<pair<Edge, Junction>> candidates;
+  m_roadGraph->FindClosestEdges(point, kMaxRoadCandidates, candidates);
+
+  double minDistance = numeric_limits<double>::max();
+  size_t minIndex = candidates.size();
+
+  for (size_t i = 0; i < candidates.size(); ++i)
+  {
+    Edge const & edge = candidates[i].first;
+    m2::DistanceToLineSquare<m2::PointD> squareDistance;
+    squareDistance.SetBounds(edge.GetStartJunction().GetPoint(), edge.GetEndJunction().GetPoint());
+    double const distance = squareDistance(point);
+    if (distance < minDistance)
+    {
+      minDistance = distance;
+      minIndex = i;
+    }
+  }
+
+  if (minIndex < candidates.size())
+  {
+    closestEdge = candidates[minIndex].first;
+    return true;
+  }
+
+  return false;
+}
+
+bool AStarRouter::LoadIndex(MwmSet::MwmId const & mwmId, IndexGraph & graph)
+{
+  MwmSet::MwmHandle mwmHandle = m_index.GetMwmHandleById(mwmId);
+  MwmValue const * mwmValue = mwmHandle.GetValue<MwmValue>();
+  if (!mwmValue)
+  {
+    LOG(LERROR, ("mwmValue == null"));
```

Comment:
So what? It's not an error - the user may remove the map during routing construction.